src/marks/raster.js

```diff
@@ -241,52 +241,34 @@ function interpolateNone(index, width, height, X, Y, V) {
   return W;
 }
 
-const ex = Symbol("extrapolate");
-
 function interpolateBarycentric(index, width, height, X, Y, V) {
   const random = randomLcg(42); // TODO allow configurable rng?
 
-  // renumber/reindex everything because we’re going to add points
-  index = Array.from(index);
-  let i = index.length;
-  X = take(X, index);
-  Y = take(Y, index);
-  V = take(V, index);
-
-  // to extrapolate, we need to fill the rectangle; pad the perimeter with vertices all around
-  const addPoint = (x, y) => ((X[i] = x), (Y[i] = y), (V[i] = ex), index.push(i++));
-  for (let j = 0, m = (width >> 2) - 1; j <= m; ++j) {
-    const k = j / m;
-    addPoint(k * width, -1);
-    addPoint((1 - k) * width, height + 1);
-  }
-  for (let j = 0, m = (height >> 2) - 1; j <= m; ++j) {
-    const k = j / m;
-    addPoint(-1, k * height);
-    addPoint(width + 1, (1 - k) * height);
-  }
+  // Flatten the input coordinates to prepare to insert extrapolated points
+  // along the perimeter of the grid (so there’s no blank output).
+  const n = index.length;
+  const nw = width >> 2;
+  const nh = (height >> 2) - 1;
+  const m = n + nw * 2 + nh * 2;
+  const XY = new Float64Array(m * 2);
+  for (let i = 0; i < n; ++i) (XY[i * 2] = X[i]), (XY[i * 2 + 1] = Y[i]);
 
-  const {points, triangles} = Delaunay.from(
-    index,
-    (i) => X[i],
-    (i) => Y[i]
-  );
+  // Add points along each edge, making sure to include the four corners for
+  // complete coverage (with no chamfered edges).
+  let i = n;
+  const addPoint = (x, y) => ((XY[i * 2] = x), (XY[i * 2 + 1] = y), i++);
+  for (let j = 0; j <= nw; ++j) addPoint((j / nw) * width, 0), addPoint((j / nw) * width, height);
+  for (let j = 0; j < nh; ++j) addPoint(width, (j / nh) * height), addPoint(0, (j / nh) * height);
 
-  // Some triangles have one undefined value; other triangles have two. Fill
-  // each undefined vertex with an average of the other defined vertices.
-  const mix2 = mixer2(V, random);
-  for (let i = 0; i < triangles.length; i += 3) {
-    const a = triangles[i];
-    const b = triangles[i + 1];
-    const c = triangles[i + 2];
-    if (V[a] === ex) V[a] = V[c] === ex ? V[b] : V[b] === ex ? V[c] : mix2(V[c], V[b]);
-    if (V[b] === ex) V[b] = V[a] === ex ? V[c] : V[c] === ex ? V[a] : mix2(V[a], V[c]);
-    if (V[c] === ex) V[c] = V[b] === ex ? V[a] : V[a] === ex ? V[b] : mix2(V[b], V[a]);
-  }
+  // To each edge point, assign the closest (non-extrapolated) value.
```

Comment:
nice!!

Reply:
Looks like this changes the extrapolation in the penguins barycentric image quite a bit, and it takes a moment to understand why there are blue triangles on the left and at the bottom-right corner. But it makes sense, and the previous version was already quite arbitrary (tossing a coin to decide on the color…).

I'm glad we have the unit test, because it shows how the algorithm operates, but it's not a good example (random-walk is the best for this task).

Reply:
Yes, I thought this was weird at first too, but then I checked the Voronoi diagram which explains it:

<img width="651" alt="Screenshot 2023-01-08 at 10 39 33 AM" src="https://user-images.githubusercontent.com/230541/211213072-2e2280fa-ae7c-4a3b-bb3c-b6de79be4444.png">

👍